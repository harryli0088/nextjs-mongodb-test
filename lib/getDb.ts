import clientPromise from "./mongodb";

export default async function getDb() {
  const client = await clientPromise;
  return client.db("test");
}