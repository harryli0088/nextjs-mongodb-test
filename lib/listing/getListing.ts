import { ListingInterface, ListingStatusType } from '../../types/listing'
import { ObjectId } from 'mongodb';
import publicizeListing from './publicizeListing';
import getDb from '../getDb';

export default async function getListing(_id:string, userId:string|null, status?:ListingStatusType) {
  const db = await getDb()

  const query = {
    _id: new ObjectId(_id),
    $or: [ { buyerId: null }, { buyerId: userId }, { sellerId: userId } ], //user can only view this listing if they are the buyer or seller, or the listing is available, ie buyerId is null
    status,
  }
  if (!status) {
    delete query.status
  }

  const listing = await db.collection("listings").findOne<ListingInterface>(query)
  if(listing) {
    publicizeListing(listing,userId)
  }

  return JSON.parse(JSON.stringify(listing)) as typeof listing
}

