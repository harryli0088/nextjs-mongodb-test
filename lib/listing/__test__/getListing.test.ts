import { ObjectId } from 'mongodb'
import { afterAll, assert, beforeAll, describe, expect, it } from 'vitest'
import getDb from '../../getDb'
import getListing from '../getListing'

describe('getListing', () => {
  let insertedIds:{[idx:number]: ObjectId} = {}
  beforeAll(async () => {
    const db = await getDb()
    const result = await db.collection("listings").insertMany(TEST_DATA);
    insertedIds = result.insertedIds
  })

  afterAll(async () => {
    const db = await getDb()
    await db.collection('listings').deleteMany({});
  })


  it('returns the appropriate listings for the seller', async () => {
    expect(await getListing(insertedIds[0].toString(), "mySellerId")).toMatchObject({
      ...TEST_DATA[0],
    })
    expect(await getListing(insertedIds[1].toString(), "mySellerId")).toMatchObject({
      ...TEST_DATA[1],
      buyerId: null,
    })
    expect(await getListing(insertedIds[2].toString(), "mySellerId")).toMatchObject({
      ...TEST_DATA[2],
      buyerId: null,
    })
    expect(await getListing(insertedIds[3].toString(), "mySellerId")).toMatchObject({
      ...TEST_DATA[3],
      buyerId: null,
    })
    expect(await getListing(insertedIds[4].toString(), "mySellerId")).toEqual(null)
    expect(await getListing(insertedIds[5].toString(), "mySellerId")).toEqual(null)
  })

  it('returns the appropriate listings for the buyer', async () => {
    expect(await getListing(insertedIds[0].toString(), "myBuyerId")).toMatchObject({
      ...TEST_DATA[0],
      sellerId: null,
    })
    expect(await getListing(insertedIds[1].toString(), "myBuyerId")).toMatchObject({
      ...TEST_DATA[1],
      sellerId: null,
    })
    expect(await getListing(insertedIds[2].toString(), "myBuyerId")).toMatchObject({
      ...TEST_DATA[2],
      sellerId: null,
    })
    expect(await getListing(insertedIds[3].toString(), "myBuyerId")).toEqual(null)
    expect(await getListing(insertedIds[4].toString(), "myBuyerId")).toMatchObject({
      ...TEST_DATA[4],
      sellerId: null,
    })
    expect(await getListing(insertedIds[5].toString(), "myBuyerId")).toEqual(null)
  })

  it('returns the appropriate listings for the a non logged in user', async () => {
    expect(await getListing(insertedIds[0].toString(), null)).toMatchObject({
      ...TEST_DATA[0],
      sellerId: null,
    })
    expect(await getListing(insertedIds[1].toString(), null)).toEqual(null)
    expect(await getListing(insertedIds[2].toString(), null)).toEqual(null)
    expect(await getListing(insertedIds[3].toString(), null)).toEqual(null)
    expect(await getListing(insertedIds[4].toString(), null)).toEqual(null)
    expect(await getListing(insertedIds[5].toString(), null)).toEqual(null)
  })

  it('filters listings for the seller by status', async () => {
    expect(await getListing(insertedIds[0].toString(), "mySellerId", "available")).toMatchObject({
      ...TEST_DATA[0],
    })
    expect(await getListing(insertedIds[1].toString(), "mySellerId", "available")).toEqual(null)
    expect(await getListing(insertedIds[2].toString(), "mySellerId", "available")).toEqual(null)
    expect(await getListing(insertedIds[3].toString(), "mySellerId", "available")).toEqual(null)
    expect(await getListing(insertedIds[4].toString(), "mySellerId", "available")).toEqual(null)
    expect(await getListing(insertedIds[5].toString(), "mySellerId", "available")).toEqual(null)
  })
})

const TEST_DATA = [
  { //available listing, mySellerId, no buyer
    buyerId: null,
    description: "my description",
    history: [{
      date: new Date().getTime(),
      description: "created",
    }],
    price: 123,
    sellerId: "mySellerId",
    status: "available",
    title: "my title",
  },
  { //in process listing, mySellerId, myBuyerId
    buyerId: "myBuyerId",
    description: "my description",
    history: [{
      date: new Date().getTime(),
      description: "created",
    }], //TODO history
    price: 123,
    sellerId: "mySellerId",
    status: "shipped",
    title: "my title",
  },
  { //completed listing, mySellerId, myBuyerId
    buyerId: "myBuyerId",
    description: "my description",
    history: [{
      date: new Date().getTime(),
      description: "created",
    }], //TODO history
    price: 123,
    sellerId: "mySellerId",
    status: "completed",
    title: "my title",
  },
  { //completed listing, mySellerId, myBuyerId2
    buyerId: "myBuyerId2",
    description: "my description",
    history: [{
      date: new Date().getTime(),
      description: "created",
    }], //TODO history
    price: 123,
    sellerId: "mySellerId",
    status: "completed",
    title: "my title",
  },
  { //completed listing, mySellerId2, myBuyerId
    buyerId: "myBuyerId",
    description: "my description",
    history: [{
      date: new Date().getTime(),
      description: "created",
    }], //TODO history
    price: 123,
    sellerId: "mySellerId2",
    status: "completed",
    title: "my title",
  },
  { //completed listing, mySellerId2, myBuyerId2
    buyerId: "myBuyerId2",
    description: "my description",
    history: [{
      date: new Date().getTime(),
      description: "created",
    }], //TODO history
    price: 123,
    sellerId: "mySellerId2",
    status: "completed",
    title: "my title",
  },
]