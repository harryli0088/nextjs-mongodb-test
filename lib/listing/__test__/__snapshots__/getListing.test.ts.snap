// Vitest Snapshot v1

exports[`getListing > snapshot 1`] = `
{
  "foo": "bar",
}
`;
